Power9: mumps does not install with XL compiler

## Summary

While building `trilinos` with `netlib-lapack`, Spack crashes while installing `mumps` with the error

```
/usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../lib64/crt1.o:(.rodata+0x8): undefined reference to `main'
```

### Expected Result

Successful `mumps` install.

### Actual Result
```
$ spack install trilinos % xl@13.1 ^netlib-lapack
==> bzip2 is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/bzip2-1.0.6-5hy4jatd26u3y233xprkhmic4ut5ryza
==> zlib is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/zlib-1.2.11-bitmk7qe3m2nl2frj6nqwhwoaat7jgat
==> boost is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/boost-1.65.1-wnwzorgwlouj4xcz2icj3r7cm6prqdrn
==> pkg-config@0.29.2 : externally installed in /usr/bin
==> pkg-config@0.29.2 : already registered in DB
==> ncurses is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/ncurses-6.0-x7esgz7hjxka474w6niigijxyzrllzje
==> openssl@1.0.2k : externally installed in /usr/bin
==> openssl@1.0.2k : already registered in DB
==> cmake is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/cmake-3.9.4-o7elw6dn22jrdomftgmo4c26eonergja
==> glm is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/glm-0.9.7.1-zbym6vapfelg65u5npwzlvufr4bykibt
==> mpich is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/mpich-3.2-x73h4dydfoybslpvsowms6m4mutbiwjq
==> hdf5 is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/hdf5-1.10.1-ctphyxffmjmffv6uyzxa66nnjknmwn7o
==> netlib-lapack is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/netlib-lapack-3.6.1-b5ncamkl5hv7hoxhexxfkjj3xrwyiz4p
==> hypre is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/hypre-2.12.1-u4lxig4om57qpnlxm6bxgfc5ocnwtksh
==> matio is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/matio-1.5.9-zn6qiqcnf56fhjvtgjdxwlfs7gedkez6
==> metis is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/metis-5.1.0-y7kgydjfcnczqt3xcpzn3cclzse5umix
==> netlib-scalapack is already installed in /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/netlib-scalapack-2.0.2-pnmwl7fyuq3n7dadgdbp46l4ocbxxbt2
==> Installing mumps
==> Fetching file:///usr/projects/draco/vendors/spack.mirror/mumps/mumps-5.1.1.tar.gz
######################################################################## 100.0%
==> Staging archive: /projects/draco/vendors/spack.power9-0.11.2/var/spack/stage/mumps-5.1.1-mor5himkwldgs4f6b5kasxgo76dw72vc/MUMPS_5.1.1.tar.gz
==> Created stage in /projects/draco/vendors/spack.power9-0.11.2/var/spack/stage/mumps-5.1.1-mor5himkwldgs4f6b5kasxgo76dw72vc
==> No patches needed for mumps
==> Building mumps [Package]
==> Executing phase: 'install'
==> Error: ProcessError: Command exited with status 2:
    'make' 'sexamples' 'cexamples' 'dexamples' 'zexamples'

4 errors found in build log:
     [ ... ]
     25    /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/mpich-3.2-x73h4dydfoybslpvsowms6m4mutbiwjq/bin/mpicc -I../include -qpic -O3  -c interface.c -ointerface.o
     26        1500-036: (I) The NOSTRICT option (default at OPT(3)) has the potential to alter the semantics of a program.  Please refer to documentation on the STRICT/NOSTRICT option for more information.
     27    /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/mpich-3.2-x73h4dydfoybslpvsowms6m4mutbiwjq/bin/mpicc -I../include -qpic -O3  -c sort.c -osort.o
     28    /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/mpich-3.2-x73h4dydfoybslpvsowms6m4mutbiwjq/bin/mpicc -I../include -qpic -O3  -c minpriority.c -ominpriority.o
     29        1500-036: (I) The NOSTRICT option (default at OPT(3)) has the potential to alter the semantics of a program.  Please refer to documentation on the STRICT/NOSTRICT option for more information.
     30    /projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/mpich-3.2-x73h4dydfoybslpvsowms6m4mutbiwjq/bin/mpif90 -shared -Wl,-soname -Wl,/projects/draco/vendors/spack.power9-0.11.2/opt/spack/linux-rhel7-ppc64le/xl-13.1/mumps-5.1.1-mor5himkwldgs4f6b5kasxgo76dw72vc/lib/libpord.so -olibpord.so graph.o gbipart.o gbisect.o ddcreate.o ddbisect.o nestdiss.o multisector.o gelim.o bucket.o tree.o symbfac.o interface.o sort.o minpriority.o
  >> 31    /usr/lib/gcc/ppc64le-redhat-linux/4.8.5/../../../../lib64/crt1.o:(.rodata+0x8): undefined reference to `main'
  >> 32    make[2]: *** [libpord.so] Error 1
     33    make[2]: Leaving directory `/scratch/users/dantopa/spack-stage/spack-stage-Yz62Av/MUMPS_5.1.1/PORD/lib'
  >> 34    make[1]: *** [lib/libpord.so] Error 2
     35    make[1]: Leaving directory `/scratch/users/dantopa/spack-stage/spack-stage-Yz62Av/MUMPS_5.1.1'
  >> 36    make: *** [s] Error 2

See build log for details:
  /projects/draco/vendors/spack.power9-0.11.2/var/spack/stage/mumps-5.1.1-mor5himkwldgs4f6b5kasxgo76dw72vc/MUMPS_5.1.1/spack-build.out
```
[spack-build.out.txt](https://github.com/spack/spack/files/1745969/spack-build.out.txt)
[spack-cc-mumps-mor5him.in.log](https://github.com/spack/spack/files/1745970/spack-cc-mumps-mor5him.in.log)
[spack-cc-mumps-mor5him.out.log](https://github.com/spack/spack/files/1745971/spack-cc-mumps-mor5him.out.log)

### Steps to reproduce the issue

```console
$ spack install trilinos % xl@13.1 ^netlib-lapack
```

### System information

 1. Platform: Darwin @ LANL
 1. Nodes: Power9
 1. Spack release 0.11.2

```
$ uname -a
Linux cn2022 4.11.0-44.4.1.el7a.ppc64le #1 SMP Sat Dec 9 02:17:51 EST 2017 ppc64le ppc64le ppc64le GNU/Linux
```

```
$ spack arch
linux-rhel7-ppc64le
```
```
$ xlc --version
IBM XL C/C++ for Linux, V13.1.6 (Community Edition)
Version: 13.01.0006.0001
/opt/ibm/xlC/13.1.6/bin/.orig/xlc: note: XL C/C++ Community Edition is a no-charge product and does not include official IBM support. You can provide feedback at the XL on POWER C/C++ Community Edition forum (http://ibm.biz/xlcpp-linux-ce). For information about a fully supported XL C/C++ compiler, visit XL C/C++ for Linux (http://ibm.biz/xlcpp-linux).

$ xlf --version
The license for the ESP version of IBM XL Fortran for Linux, V16.1 (Beta) compiler product has expired.
```
Configuration files:
[compilers.config.txt](https://github.com/spack/spack/files/1745851/compilers.config.txt)
[config.config.txt](https://github.com/spack/spack/files/1745852/config.config.txt)
[modules.config.txt](https://github.com/spack/spack/files/1745853/modules.config.txt)
[packages.config.txt](https://github.com/spack/spack/files/1745854/packages.config.txt)

--- a/var/spack/repos/builtin/packages/mumps/package.py
+++ b/var/spack/repos/builtin/packages/mumps/package.py
@@ -232,10 +232,18 @@ class Mumps(Package):
             else:
                 makefile_conf.extend([
                     'LIBEXT=.so',
-                    'AR=$(FL) -shared -Wl,-soname -Wl,%s/$(notdir $@) %s -o' %
+                    'AR=link_cmd() { $(FL) -shared -Wl,-soname '
+                    '-Wl,%s/$(notdir $@) -o "$$@" %s; }; link_cmd ' %
                     (prefix.lib, inject_libs),
-                    'RANLIB=echo'
+                    'RANLIB=ls'
                 ])
+                # When building libpord, read AR from Makefile.inc instead of
+                # going through the make command line - this prevents various
+                # problems with the substring "$$@".
+                filter_file(' AR="\$\(AR\)"', '', 'Makefile')
+                filter_file('^(INCLUDES = -I../include)',
+                            '\\1\ninclude ../../Makefile.inc',
+                            join_path('PORD', 'lib', 'Makefile'))
 
                 if using_xl:
                     # The patches for xl + spectrum-mpi use SAR for linking

--- a/var/spack/repos/builtin/packages/mumps/package.py
+++ b/var/spack/repos/builtin/packages/mumps/package.py
@@ -203,10 +203,7 @@ class Mumps(Package):
 
         if '+shared' in self.spec:
             # All Mumps libraries will be linked with 'inject_libs'.
-            # Usually, the rpaths will be injected by the Spack compiler
-            # wrapper, however some MPI wrappers may not call the Spack
-            # compiler wrapper.
-            inject_libs = [self.rpath_args]
+            inject_libs = []
             if '+mpi' in self.spec:
                 inject_libs += [self.spec['scalapack'].libs.ld_flags]
             if '+ptscotch' in self.spec or '+scotch' in self.spec:

--- a/var/spack/repos/builtin/packages/mumps/package.py
+++ b/var/spack/repos/builtin/packages/mumps/package.py
@@ -85,16 +85,26 @@ class Mumps(Package):
             raise RuntimeError(
                 'You cannot use the variants parmetis or ptscotch without mpi')
 
-        lapack_blas = (self.spec['lapack'].libs +
-                       self.spec['blas'].libs)
-        makefile_conf = ["LIBBLAS = %s" % lapack_blas.ld_flags]
+        # The makefile variables LIBBLAS, LSCOTCH, LMETIS, and SCALAP are only
+        # used to link the examples, so if building '+shared' there is no need
+        # to explicitly link with the respective libraries because we make sure
+        # the mumps shared libraries are already linked with them. See also the
+        # comment below about 'inject_libs'. This behaviour may cause problems
+        # if building '+shared' and the used libraries were build static
+        # without the PIC option.
+        shared = '+shared' in self.spec
+
+        lapack_blas = (self.spec['lapack'].libs + self.spec['blas'].libs)
+        makefile_conf = ["LIBBLAS = %s" %
+                         lapack_blas.ld_flags if not shared else '']
 
         orderings = ['-Dpord']
 
         if '+ptscotch' in self.spec or '+scotch' in self.spec:
             makefile_conf.extend([
                 "ISCOTCH = -I%s" % self.spec['scotch'].prefix.include,
-                "LSCOTCH = {0}".format(self.spec['scotch'].libs.ld_flags)
+                "LSCOTCH = {0}".format(
+                    self.spec['scotch'].libs.ld_flags if not shared else '')
             ])
 
             orderings.append('-Dscotch')
@@ -104,16 +114,19 @@ class Mumps(Package):
         if '+parmetis' in self.spec and '+metis' in self.spec:
             makefile_conf.extend([
                 "IMETIS = -I%s" % self.spec['parmetis'].prefix.include,
-                "LMETIS = -L%s -l%s -L%s -l%s" % (
+                ("LMETIS = -L%s -l%s -L%s -l%s" % (
                     self.spec['parmetis'].prefix.lib, 'parmetis',
-                    self.spec['metis'].prefix.lib, 'metis')
+                    self.spec['metis'].prefix.lib, 'metis')) if not shared
+                else 'LMETIS ='
             ])
 
             orderings.append('-Dparmetis')
         elif '+metis' in self.spec:
             makefile_conf.extend([
                 "IMETIS = -I%s" % self.spec['metis'].prefix.include,
-                "LMETIS = -L%s -l%s" % (self.spec['metis'].prefix.lib, 'metis')
+                ("LMETIS = -L%s -l%s" % (
+                    self.spec['metis'].prefix.lib, 'metis')) if not shared
+                else 'LMETIS ='
             ])
 
             orderings.append('-Dmetis')
@@ -127,6 +140,7 @@ class Mumps(Package):
         # TODO: test this part, it needs a full blas, scalapack and
         # partitionning environment with 64bit integers
 
+        using_xl = self.compiler.name in ['xl', 'xl_r']
         if '+int64' in self.spec:
             if self.compiler.name == "xl" or self.compiler.name == "xl_r":
                 makefile_conf.extend(
@@ -142,7 +156,7 @@ class Mumps(Package):
                      'OPTL    = %s -O ' % fpic,
                      'OPTC    = %s -O -DINTSIZE64' % fpic])
         else:
-            if self.compiler.name == "xl" or self.compiler.name == "xl_r":
+            if using_xl:
                 makefile_conf.extend(
                     ['OPTF    = -O3 -qfixed',
                      'OPTL    = %s -O3' % fpic,
@@ -154,13 +168,16 @@ class Mumps(Package):
                      'OPTC    = %s -O ' % fpic])
 
         if '+mpi' in self.spec:
-            scalapack = self.spec['scalapack'].libs
+            scalapack = self.spec['scalapack'].libs if not shared \
+                else LibraryList([])
             makefile_conf.extend(
                 ['CC = {0}'.format(self.spec['mpi'].mpicc),
                  'FC = {0}'.format(self.spec['mpi'].mpifc),
                  "SCALAP = %s" % scalapack.ld_flags,
                  "MUMPS_TYPE = par"])
-            if (self.spec.satisfies('%xl_r' or '%xl')) and self.spec.satisfies('^spectrum-mpi'):  # noqa
+            # The FL makefile variable is used for linking the examples and
+            # linking the shared mumps libraries (in some cases).
+            if using_xl and self.spec.satisfies('^spectrum-mpi'):
                 makefile_conf.extend(
                     ['FL = {0}'.format(self.spec['mpi'].mpicc)])
             else:
@@ -181,28 +198,54 @@ class Mumps(Package):
             # hack defined by _DMAIN_COMP (see examples/c_example.c)
             makefile_conf.append("CDEFS   = -DAdd_ -DMAIN_COMP")
         else:
-            if self.compiler.name != "xl" and self.compiler.name != "xl_r":
+            if not using_xl:
                 makefile_conf.append("CDEFS   = -DAdd_")
 
         if '+shared' in self.spec:
+            # All Mumps libraries will be linked with 'inject_libs'.
+            # Usually, the rpaths will be injected by the Spack compiler
+            # wrapper, however some MPI wrappers may not call the Spack
+            # compiler wrapper.
+            inject_libs = [self.rpath_args]
+            if '+mpi' in self.spec:
+                inject_libs += [self.spec['scalapack'].libs.ld_flags]
+            if '+ptscotch' in self.spec or '+scotch' in self.spec:
+                inject_libs += [self.spec['scotch'].libs.ld_flags]
+            if '+parmetis' in self.spec and '+metis' in self.spec:
+                inject_libs += [
+                    "-L%s -l%s -L%s -l%s" % (
+                        self.spec['parmetis'].prefix.lib, 'parmetis',
+                        self.spec['metis'].prefix.lib, 'metis')]
+            elif '+metis' in self.spec:
+                inject_libs += [
+                    "-L%s -l%s" % (self.spec['metis'].prefix.lib, 'metis')]
+            inject_libs += [lapack_blas.ld_flags]
+            inject_libs = ' '.join(inject_libs)
+
             if sys.platform == 'darwin':
                 # Building dylibs with mpif90 causes segfaults on 10.8 and
                 # 10.10. Use gfortran. (Homebrew)
                 makefile_conf.extend([
                     'LIBEXT=.dylib',
-                    'AR=%s -dynamiclib -Wl,-install_name -Wl,%s/$(notdir $@) -undefined dynamic_lookup -o ' % (os.environ['FC'], prefix.lib),  # noqa
+                    'AR=%s -dynamiclib -Wl,-install_name -Wl,%s/$(notdir $@)'
+                    ' -undefined dynamic_lookup %s -o ' %
+                    (os.environ['FC'], prefix.lib, inject_libs),
                     'RANLIB=echo'
                 ])
             else:
                 makefile_conf.extend([
                     'LIBEXT=.so',
-                    'AR=$(FL) -shared -Wl,-soname -Wl,%s/$(notdir $@) -o' % prefix.lib,  # noqa
+                    'AR=$(FL) -shared -Wl,-soname -Wl,%s/$(notdir $@) %s -o' %
+                    (prefix.lib, inject_libs),
                     'RANLIB=echo'
                 ])
 
-                if self.compiler.name == 'xl' or self.compiler.name == 'xl_r':
+                if using_xl:
+                    # The patches for xl + spectrum-mpi use SAR for linking
+                    # libpord.
                     makefile_conf.extend([
-                        'SAR=/bin/xlc -shared -Wl,-soname -Wl,%s/$(notdir $@) -o' % prefix.lib  # noqa
+                        'SAR=%s -shared -Wl,-soname -Wl,%s/$(notdir $@) %s -o'
+                        % (env['CC'], prefix.lib, inject_libs)
                     ])
         else:
             makefile_conf.extend([
@@ -222,24 +265,17 @@ class Mumps(Package):
                 fh.write(makefile_inc)
 
     def install(self, spec, prefix):
-        make_libs = []
-
-        # the choice to compile ?examples is to have kind of a sanity
-        # check on the libraries generated.
-        if '+float' in spec:
-            make_libs.append('sexamples')
-            if '+complex' in spec:
-                make_libs.append('cexamples')
-
-        if '+double' in spec:
-            make_libs.append('dexamples')
-            if '+complex' in spec:
-                make_libs.append('zexamples')
-
         self.write_makefile_inc()
 
         # Build fails in parallel
-        make(*make_libs, parallel=False)
+        # That is why we split the builds of 's', 'c', 'd', and/or 'z' which
+        # can be build one after the other, each using a parallel build.
+        letters_variants = [
+            ['s', '+float'], ['c', '+complex+float'],
+            ['d', '+double'], ['z', '+complex+double']]
+        for l, v in letters_variants:
+            if v in spec:
+                make(l + 'examples')
 
         install_tree('lib', prefix.lib)
         install_tree('include', prefix.include)
@@ -253,15 +289,23 @@ class Mumps(Package):
 
         # FIXME: extend the tests to mpirun -np 2 when build with MPI
         # FIXME: use something like numdiff to compare output files
-        with working_dir('examples'):
-            if '+float' in spec:
-                os.system('./ssimpletest < input_simpletest_real')
-                if '+complex' in spec:
-                    os.system('./csimpletest < input_simpletest_real')
-            if '+double' in spec:
-                os.system('./dsimpletest < input_simpletest_real')
-                if '+complex' in spec:
-                    os.system('./zsimpletest < input_simpletest_cmplx')
+        # Note: In some cases, when 'mpi' is enabled, the examples below cannot
+        # be run without 'mpirun', so we enabled the tests only if explicitly
+        # requested with the Spack '--test' option.
+        if self.run_tests:
+            with working_dir('examples'):
+                if '+float' in spec:
+                    ssimpletest = Executable('./ssimpletest')
+                    ssimpletest(input='input_simpletest_real')
+                    if '+complex' in spec:
+                        csimpletest = Executable('./csimpletest')
+                        csimpletest(input='input_simpletest_cmplx')
+                if '+double' in spec:
+                    dsimpletest = Executable('./dsimpletest')
+                    dsimpletest(input='input_simpletest_real')
+                    if '+complex' in spec:
+                        zsimpletest = Executable('./zsimpletest')
+                        zsimpletest(input='input_simpletest_cmplx')
 
     @property
     def libs(self):

